File: algo/238.py
from typing import List

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        out = []
        p = 1
        for i in range(len(nums)):
            out.append(p)
            p *= nums[i]
        
        p = 1
        for i in range(len(nums) - 1, -1 , -1):
            out[i] *= p
            p *= nums[i]
        return out


nums = [1,2,3,4]
print(Solution().productExceptSelf(nums))
nums = [-1,1,0,-3,3]
print(Solution().productExceptSelf(nums))

File: algo/238.kt
class Solution {
    fun productExceptSelf(nums: IntArray): IntArray {
        var p = 1
        var ans = IntArray(nums.size)
        ans[0] = p
        for (i in 0..nums.size - 2) {
            p *= nums[i]
            ans[i + 1] = p
        }

        p = 1
        for (i in nums.size - 1 downTo 1) {
            p *= nums[i]
            ans[i - 1] *= p
        }
        return ans
    }
}
